File: src/main/java/at/yawk/math/data/GeneratedVariableExpression.java
package at.yawk.math.data;

import java.util.UUID;
import lombok.EqualsAndHashCode;

/**
 * @author yawkat
 */
@EqualsAndHashCode
public class GeneratedVariableExpression implements Expression {
    private final UUID id = UUID.randomUUID();

    @Override
    public String toString(int radix) {
        return "v[" + id + ']';
    }
}

File: src/main/kotlin/at/yawk/math/data/GeneratedVariableExpression.kt
package at.yawk.math.data

import at.yawk.math.EqualsHelper
import java.util.*

/**
 * @author yawkat
 */
class GeneratedVariableExpression : BaseExpression() {
    private val id = UUID.randomUUID()

    override fun toString(radix: Int): String {
        return "v[$id]"
    }

    override fun equals(other: Any?): Boolean {
        return EqualsHelper.equals<GeneratedVariableExpression>(other, { it.id == id })
    }

    override fun hashCode(): Int {
        return EqualsHelper.hashCode(id)
    }

    override fun visit(visitor: ExpressionVisitor): Expression {
        return visitor.visitSingleExpression(this)
    }
}